app/routers/sessions.py

```diff
@@ -78,30 +96,54 @@ async def create_session(session: Session):
     return JSONResponse(status_code=status.HTTP_201_CREATED, content=created_session)
 
 
-@router.patch("/{session_id}", response_model=SessionResponse)
-async def update_session(session_id: str, session: UpdateSession):
-    session = jsonable_encoder(session)
-
-    if "has_quiz_ended" not in session:
-        raise HTTPException(
-            status_code=status.HTTP_400_BAD_REQUEST,
-            detail="No value provided for 'has_quiz_ended'",
-        )
-
-    if (client.quiz.sessions.find_one({"_id": session_id})) is None:
+@router.patch("/{session_id}", response_model=UpdateSessionResponse)
+async def update_session(session_id: str, session_updates: UpdateSession):
+    """
+    session is updated whenever start/resume button is clicked
+    (or) end button is clicked for the first time
+    """
+    session_updates = jsonable_encoder(session_updates)
+    session = client.quiz.sessions.find_one({"_id": session_id})
+    if session is None:
         raise HTTPException(
             status_code=status.HTTP_404_NOT_FOUND,
             detail=f"session {session_id} not found",
         )
 
-    # update the document in the sessions collection
-    client.quiz.sessions.update_one({"_id": session_id}, {"$set": session})
+    current_time = datetime.utcnow()
+    time_elapsed = 0
+
+    if session_updates["has_quiz_started_first_time"]:
+        # start/resume button clicked first time for a quiz
+        session["has_quiz_started"] = True
+        # time elapsed is zero
+    else:
+        # update time remaining based on current time (resume/end button clicked)
+        time_elapsed = (
+            current_time - datetime.fromisoformat(session["quiz_start_resume_time"])
+        ).seconds
```

Comment:
this part is a little unclear to me.
My best guess after reading this and frontend's code is that the `else` condition is for the quiz being resumed after being paused right?

Just thinking out loud, I'll write my understanding process of this logic below.
- User comes to a quiz for the first time, a new session is created
- session['time_remaining'] is set
- user clicks "Start Quiz"
- a call to update_session API is made
- in the logic we check if the quiz is being started, resumed or ended
- if the quiz is being started and NOT resumed, we set the `quiz_start_time` in the session as the current time. time elapsed would be 0 as the quiz is just being started. so we calculate the time remaining and send it to the frontend and also update in the session document
- lets say after some time user leaves the quiz.
- they come back again, a new session is created
- this new session has details copied from the previous session. things like `has quiz started/resumed/ended` and the time remaining, quiz start time, quiz end time and quiz resume time(s). These details are outdated as they'll get updated when the user clicks resume etc.
- The user clicks resume, a call to update session API is called.
- this time the quiz is being resumed. so we'll calculate the time that has been passed since the last time session was resumed. if this is the first time session is being resumed, we'll calculate the time passed from the quiz start time
- now we have the time elapsed since the last activity (it could be quiz started or resumed)
- we'll calculate how much time is remaining since the last time_remaining and update it on DB and frontend

While writing this out the process became clear to me. Your logic is totally correct but I think it can be structured more neatly and in a way that's extendable in the future. Writing my points below so I can discuss with you on a call.

- We should segregate events. Think of starting, resuming, ending as events. Currently in your implementation, things are mixed up. ending is a separate event. starting/resuming is merged into one field. plus we're relying on the frontend to send data telling the backend that if the user is starting the quiz for the first time or not or ending the quiz for the first time or not. these are very specific events.
- Plus I think every session should contain an event string. This could be (for now) "quiz_start", "quiz_resume" and "quiz_end". Along with each event, there is a timestamp attached. My goal here is to reuse the `created_at` and `updated_at` timestamps instead of storing times in separate variables like `quiz_start_time`, `quiz_resume_time` and `quiz_end_time`.
- Going by the current implementation, In the future if we want to track some other event along with their times, we'll have to add specialized fields again. For example, if the event we want to track is `xyz`, we'll have to add `quiz_xyz_time` and then another flag let's say `has_quiz_xyz`. A better approach would be to have a general event. Which could be anything. 
- Every session will have one session-level event attached to it. And the time for the session-level event will be tracked using created_at/updated_at.
- That way, we won't have to keep track of _other_, events in one session itself. Like currently, for every session, we're tracking if the quiz was started, resumed or ended along with their times.
- With the new implementation, things will go along with the arrow of time. one session will have quiz started event with the time being tracked in created_at/updated_at. Next session will have quiz resumed event along with the time. then a few more quiz resumed events in the next few sessions. then at the end, we'll have quiz ended event. same way we can have quiz reattempted, revisited etc in the future if required.